pkg/provider/azure/keyvault/keyvault.go

```diff
@@ -0,0 +1,259 @@
+package keyvault
+
+import (
+	"context"
+	"crypto/x509"
+	"encoding/base64"
+	"encoding/pem"
+	"fmt"
+	"os"
+	"path"
+	"strings"
+
+	"github.com/Azure/azure-sdk-for-go/profiles/latest/keyvault/keyvault"
+	kvauth "github.com/Azure/azure-sdk-for-go/services/keyvault/auth"
+	"golang.org/x/crypto/pkcs12"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	esv1alpha1 "github.com/external-secrets/external-secrets/apis/externalsecrets/v1alpha1"
+	smmeta "github.com/external-secrets/external-secrets/apis/meta/v1"
+	"github.com/external-secrets/external-secrets/pkg/provider"
+	"github.com/external-secrets/external-secrets/pkg/provider/schema"
+)
+
+type Azure struct {
+	kube       client.Client
+	store      esv1alpha1.GenericStore
+	baseClient *keyvault.BaseClient
+	namespace  string
+	iAzure     IAzure
+}
+
+type IAzure interface {
+	getKeyVaultSecrets(ctx context.Context, vaultName string, version string, secretName string, withTags bool) (map[string][]byte, error)
+}
+
+func init() {
+	schema.Register(&Azure{}, &esv1alpha1.SecretStoreProvider{
+		AzureKV: &esv1alpha1.AzureKVProvider{},
+	})
+}
+
+func (a *Azure) New(ctx context.Context, store esv1alpha1.GenericStore, kube client.Client, namespace string) (provider.Provider, error) {
+	anAzure := &Azure{
+		kube:      kube,
+		store:     store,
+		namespace: namespace,
+	}
+	anAzure.iAzure = anAzure
+	azClient, err := anAzure.newAzureClient(ctx)
+
+	if err != nil {
+		return nil, err
+	}
+
+	anAzure.baseClient = azClient
+	return anAzure, nil
+}
+
+// implement store.Client.GetSecret Interface.
+// retrieve a secret with the secret name defined in ref.Property in a specific keyvault with the name ref.Name.
+func (a *Azure) GetSecret(ctx context.Context, ref esv1alpha1.ExternalSecretDataRemoteRef) ([]byte, error) {
+	version := ""
+	var secretBundle []byte
+
+	if ref.Version != "" {
+		version = ref.Version
+	}
+	secretName := ref.Property
+	nameSplitted := strings.Split(secretName, "_")
+	getTags := false
+
+	if nameSplitted[len(nameSplitted)-1] == "TAG" {
```

Comment:
The Azure KV API  is a little special: the response contains a `Value string` aswell as metadata as `Tags map[string]*string` ([docs](https://pkg.go.dev/github.com/Azure/azure-sdk-for-go/services/keyvault/v7.0/keyvault#SecretBundle)).

What's your use-case for those tags? Do you want to store multiple k/v pairs in one secret? would it make sense to push that data to the template (see below)?

Reply:
Hello @moolen , 
in our implementation , we needed the use case where the users need to reconcile the `Tags` saved within the secret in the KV , we didnt want to complicate the agnostic ExternalSecret CRD definition by adding tags that are only needed in the special case of Azure . So we thought of this solution to give this naming convention of allowing the user to grab a tag value independant from the secret value by using a documented agreed-upon naming convention by adding this suffix (same strategy we used for grabbing certificates in the rest of the code below)

Reply:
> we didnt want to complicate the agnostic ExternalSecret CRD

i see. Tho having a magic suffix `*_TAGS` / `_CERT` / `_KEY` is misleading for users (imagine having an actual secret `SECRET_TAGS`). 

I see two options for these use-cases:

(1) a second `azure-kv-tags` provider
This provider only fetches the tags from secrets as a map. The keys can be used as usual with `remoteRef.property`.

(2) Templating

The Provider may pass arbitrary `metadata` to the templating engine: 

```yaml
spec:
  secretStoreRef:
    name: my-az-store
  data:
  # note: just a single secret
  - secretKey: user
     remoteRef:
       key: my-key
  - secretKey: cert
     remoteRef:
       key: my-cert
  target:
    template:
      data:
        # .data.user is the actual secret .Value from azkv 
        # .metadata.user is metadata from azure (e.g. version, Tags, content-type)
        example.yml: |
          endpoints:
          - url: https://{{ .data.user }}:{{ .data.password }}@api.exmaple.com
             headers:
               UserAgent: {{.metadata.user.tags.USER_AGENT }}
        # different option to transform the secret value pkcs12 -> []PEM -> cert
        # this is just an idea
        mycert: '{{ .data.cert. | extractPKCS12 | filterPEMCert }}'

```

(3) actually extend the CRD

```yaml
# fetches value of tag 'baz'
remoteRef:
  key: my_secret_foo
  version: v1
  property: baz
  azureKV: # add this
    fetchTags: true # tell the provider to fetch the tags
```

I think i lean towards (3) 
What do you think?

/cc @knelasevero  @mcavoyk

Reply:
agree with you .

solution 1 may be confusing and extra work to do by implementing a totally new backend for a sake of minor feature (and honestly will not be widely used)

solution 2 can be a bit complicated

Reply:
Ok, i'll open a PR in the crd-spec repo

Reply:
I'll opened https://github.com/external-secrets/crd-spec/pull/4. Let's move the discussion there.

Reply:
@ahmedmus21 quick question regarding the tags just to be clear: Do you want to
(1) use the tags in `secret.data` ?
(2) use the tags in `secret.metadata` ?

Reply:
I will let @ahmedmus21 give a definitive answer (when back from a few well deserved days off), but given that Tags are set at Azure KeyVault object level (not secret store), and we want to have every secrets at once, I would say the secrets tags would fit in `secret.data` rather than `secret.metadata`.